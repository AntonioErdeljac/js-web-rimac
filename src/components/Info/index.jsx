import CountUp from 'react-countup';
import cn from 'classnames';
import PropTypes from 'prop-types';
import React, { memo, useEffect, useMemo } from 'react';
import { Card, FormControl } from 'react-bootstrap';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';

import './index.scss';

import { useSystem } from '../../providers';

const Info = ({ icon, title, value, suffix, min, max, warning, edit, onChange }) => {
  const { warn, clearWarning } = useSystem();

  const isFaulty = useMemo(() => {
    const maxError = value > max;
    const minError = value < min;

    return maxError || minError;
  }, [value, min, max]);

  useEffect(() => {
    if (isFaulty && !edit) {
      warn(warning);
    } else if (!edit) {
      clearWarning(warning);
    }
  }, [isFaulty, warning, warn, clearWarning, edit]);

  return (
    <Card className="r-info mb-4">
      <Card.Body className="d-flex align-items-center flex-column">
        <FontAwesomeIcon size="2x" className="r-info r-info-icon" icon={icon} />
        <div className="mt-4">
          <strong className="text-primary">{title}</strong>
        </div>
        {edit ? (
          <div className="mt-3">
            <FormControl className="my-2" placeholder="Min" onChange={onChange} name="min" />
            <FormControl placeholder="Max" onChange={onChange} name="max" />
          </div>
        ) : (
          <>
            <div className="mt-4">
              <strong className={cn({ 'text-danger': isFaulty })}>
                <CountUp preserveValue duration={5} end={value} />
                {suffix}
              </strong>
            </div>
            <hr className="w-100" />
            <div className="d-inline-flex align-items-center justify-content-between w-100 px-5">
              <small className="text-muted">Max: {max}</small>
              <small className="text-muted">Min: {min}</small>
            </div>
          </>
        )}
      </Card.Body>
    </Card>
  );
};

Info.propTypes = {
  icon: PropTypes.string.isRequired,
  title: PropTypes.string.isRequired,
  value: PropTypes.number,
  suffix: PropTypes.string,
  min: PropTypes.number,
  max: PropTypes.number,
  warning: PropTypes.string,
  edit: PropTypes.bool,
  onChange: PropTypes.func,
};

Info.defaultProps = {
  suffix: '',
  value: 0,
  min: -Infinity,
  max: Infinity,
  edit: false,
  onChange: () => {},
  warning: '',
};

export default memo(Info);
