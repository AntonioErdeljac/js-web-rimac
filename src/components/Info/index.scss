.r-info {
  border: none !important;

  &-icon {
    opacity: 0.4;
  }
}
