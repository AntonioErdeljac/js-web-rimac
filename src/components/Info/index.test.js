import React from 'react';
import toJson from 'enzyme-to-json';
import { shallow } from 'enzyme';

import Info from './index';

const mockWarn = jest.fn();
const mockClearWarning = jest.fn();

jest.mock('../../providers', () => ({
  useSystem: () => ({
    warn: mockWarn,
    clearWarning: mockClearWarning,
  }),
}));

const makeProps = ({ icon, title, value, suffix, min, max, warning, edit, onChange } = {}) => ({
  icon: icon || 'battery',
  title: title || 'Battery',
  value: value || 100,
  suffix: suffix || '%',
  min: min || 0,
  max: max || 0,
  warning: warning || 'Battery level not ideal',
  edit: edit || false,
  onChange: onChange || (() => {}),
});

const renderComponent = (props) => shallow(<Info {...props} />);

describe('Info Component', () => {
  it('renders correctly according to snapshot', () => {
    expect(toJson(renderComponent(makeProps()))).toMatchSnapshot();
  });

  it('calls warn function with warning value if value is faulty', () => {
    const warning = 'Battery level bad.';
    renderComponent(makeProps({ value: 10, min: 30, max: 100, warning }));

    expect(mockWarn.mock.calls[0][0]).toEqual(warning);
  });

  it('calls clearWarning function if value is good', () => {
    const warning = 'Battery level bad.';
    renderComponent(makeProps({ value: 50, min: 30, max: 100, warning }));

    expect(mockClearWarning).toHaveBeenCalled();
  });
});
