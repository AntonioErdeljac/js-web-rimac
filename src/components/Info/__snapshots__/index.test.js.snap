// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Info Component renders correctly according to snapshot 1`] = `
<Card
  body={false}
  className="r-info mb-4"
>
  <CardBody
    className="d-flex align-items-center flex-column"
  >
    <FontAwesomeIcon
      border={false}
      className="r-info r-info-icon"
      fixedWidth={false}
      flip={null}
      icon="battery"
      inverse={false}
      listItem={false}
      mask={null}
      pull={null}
      pulse={false}
      rotation={null}
      size="2x"
      spin={false}
      swapOpacity={false}
      symbol={false}
      title=""
      transform={null}
    />
    <div
      className="mt-4"
    >
      <strong
        className="text-primary"
      >
        Battery
      </strong>
    </div>
    <div
      className="mt-4"
    >
      <strong
        className=""
      >
        <CountUp
          decimal="."
          decimals={0}
          delay={null}
          duration={5}
          easingFn={null}
          end={100}
          formattingFn={null}
          onEnd={[Function]}
          onPauseResume={[Function]}
          onReset={[Function]}
          onStart={[Function]}
          onUpdate={[Function]}
          prefix=""
          preserveValue={true}
          redraw={false}
          separator=""
          start={0}
          startOnMount={true}
          suffix=""
          useEasing={true}
        />
        %
      </strong>
    </div>
    <hr
      className="w-100"
    />
    <div
      className="d-inline-flex align-items-center justify-content-between w-100 px-5"
    >
      <small
        className="text-muted"
      >
        Min: 
        0
      </small>
      <small
        className="text-muted"
      >
        Max: 
        0
      </small>
    </div>
  </CardBody>
</Card>
`;
